README.md
# swiftnav_piksi
Source code for the ROS driver for the Swiftnav Pksi RTK GPS receivers

This ROS driver has been tested with the Piksi Multi GNSS Module, and the driver relies on the Swift Binary Protocol library which can be obtained [here](https://github.com/swift-nav/libsbp). Note that testing was performed with v2.1.7. The installation instructions can be found [here](https://swift-nav.github.io/libsbp/c/build/docs/html/install.html). 


swiftnav_piksi/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package swiftnav_piksi
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.4 (2016-03-03)
------------------
* First public release for indigo and jade


swiftnav_piksi/src/piksi.c
/***************************************************************************//**
* \file piksi.c
*
* \brief Standalone C Driver for the Swift-Nav Piksi
* \author Scott K Logan
* \author Caleb Jamison
* \date Febuary 23, 2014
*
* This is a standolone C driver for the Swift Navigation Piksi GPS.
*
* \section license License (BSD-3)
* Copyright (c) 2013, Scott K Logan\n
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* - Redistributions of source code must retain the above copyright notice,
* this list of conditions and the following disclaimer.
* - Redistributions in binary form must reproduce the above copyright notice,
* this list of conditions and the following disclaimer in the documentation
* and/or other materials provided with the distribution.
* - Neither the name of Willow Garage, Inc. nor the names of its contributors
* may be used to endorse or promote products derived from this software
* without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
******************************************************************************/

#include "swiftnav_piksi/piksi.h"

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <fcntl.h>
#include <stdio.h>

/*!
 * \brief Maximum number of simultaneously open device handles.
 */
#define MAX_HANDLES 256

#define NO_FLUSH_BUFFER 1

struct piksi_priv
{
	int fd;
	int baud;
	char *port;
};

/*!
 * \brief List of communication handles.
 */
static struct piksi_priv * piksi_list[MAX_HANDLES] = { NULL };


/*!
 * \brief Grabs the next available device handle slot.
 *
 * Iterates through the ::MAX_HANDLES slots for the lowest available index.
 *
 * \returns Open slot index between 0 and ::MAX_HANDLES
 * \retval -1 No available slots
 */
static int next_available_handle( )
{
	unsigned short int i;
	for( i = 0; i < MAX_HANDLES; i++ )
	{
		if( !piksi_list[i] )
			return i;
	}
	return -1;
}

static int baud2term( int baud )
{
	switch( baud )
	{
	case 1200:
		return B1200;
		break;
	case 2400:
		return B2400;
		break;
	case 4800:
		return B4800;
		break;
	case 9600:
		return B9600;
		break;
	case 19200:
		return B19200;
		break;
	case 38400:
		return B38400;
		break;
	case 57600:
		return B57600;
		break;
	case 115200:
		return B115200;
		break;
	case 230400:
		return B230400;
		break;
	case 460800:
		return B460800;
		break;
	case 921600:
		return B921600;
		break;
	case 1000000:
		return B1000000;
		break;
	default:
		return B0;
		break;
	}
}

u32 send_cmd( u8 *data, u32 num_bytes, void* context )
{
	int8_t piksid = *( int8_t* ) context;
	return write( piksi_list[piksid]->fd, data, num_bytes );
}

u32 read_data( u8 *data, u32 num_bytes, void* context )
{
	int8_t piksid = *( int8_t* ) context;
	u32 bytes_recv = 0;
	bytes_recv = read( piksi_list[piksid]->fd, data, num_bytes );
	if( bytes_recv <= 0 )
		return 0;
	return bytes_recv;
}

int piksi_open( const char *port, int baud )
{

	printf("Baud %d\n",baud);
	/* Step 1: Make sure the device opens OK */
	int fd = open( port, O_RDWR | O_NOCTTY | O_NDELAY );
	if( fd < 0 )
		return PIKSI_ERROR_NO_DEVICE;

	fcntl( fd, F_SETFL, 0 );

	struct termios options;
	cfmakeraw( &options );
	if( cfsetispeed( &options, baud2term(baud) ) < 0 )
	{
		close( fd );
		return PIKSI_ERROR_IO;
	}
	options.c_cflag &= ~HUPCL;
	options.c_lflag &= ~ICANON;
	options.c_cc[VTIME] = 2;
	options.c_cc[VMIN] = 0;
	if( tcsetattr( fd, TCSANOW, &options ) < 0 )
	{
		close( fd );
		return PIKSI_ERROR_IO;
	}

	/* Step 2: Allocate a private struct */
	int mydev = next_available_handle( );
	if( mydev < 0 )
		goto close_mem_error;

	piksi_list[mydev] = malloc( sizeof( struct piksi_priv ) );
	if( !piksi_list[mydev] )
		goto null_struct;

	memset( piksi_list[mydev], 0, sizeof( struct piksi_priv ) );

	piksi_list[mydev]->port = malloc( strlen( port ) + 1 );
	if( !piksi_list[mydev]->port )
		goto free_struct;

	memcpy( piksi_list[mydev]->port, port, strlen( port ) + 1 );
	piksi_list[mydev]->fd = fd;
	piksi_list[mydev]->baud = baud2term(baud);

	return mydev;

	/* Free gotos */

	free_struct:
		free( piksi_list[mydev] );
	null_struct:	
		piksi_list[mydev] = NULL;
	close_mem_error:
		close( fd );
		return PIKSI_ERROR_NO_MEM;
}

void piksi_close( const int8_t piksid )
{
	if( piksid < 0 || piksid > MAX_HANDLES || !piksi_list[piksid] )
		return;

	close( piksi_list[piksid]->fd );

	free( piksi_list[piksid]->port );
	free( piksi_list[piksid] );
	piksi_list[piksid] = NULL;

	return;
}

int main()
{
	return 0;
}



swiftnav_piksi/src/piksi_node.cpp
/***************************************************************************//**
 * \file piksi_node.cpp
 *
 * \brief Single GPS node
 * \author Scott K Logan
 * \author Caleb Jamison
 * \date February 23, 2014
 *
 * This binary creates a simple node for communication with a single Swift
 * Navigation Piksi GPS.
 *
 * \section license License (BSD-3)
 * Copyright (c) 2013, Scott K Logan\n
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * - Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of Willow Garage, Inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software
 * without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

#include <swiftnav_piksi/piksi_driver.h>

#include <ros/ros.h>
#include <cstdlib>

/*!
* \brief Main Function
*
* \author Scott K Logan
*
* Initializes ROS, instantiates the node handle for the driver to use and
* instantiates the PIKSI class.
*
* \param argc Number of command line arguments
* \param argv 2D character array of command line arguments
*
* \returns EXIT_SUCCESS, or an error state
*/
int main( int argc, char *argv[] )
{
	ros::init( argc, argv, "piksi_node" );

	ros::NodeHandle nh;
	ros::NodeHandle nh_priv( "~" );

	std::string port;
	nh_priv.param( "port", port, (const std::string)"/dev/ttyUSB0" );
	int baud;
	nh_priv.param( "baud", baud, 152000 );

	swiftnav_piksi::PIKSI piksi( nh, nh_priv, port );

	ROS_DEBUG( "Opening Piksi on %s", port.c_str( ) );
	if( !piksi.PIKSIOpen( ) )
		ROS_ERROR( "Failed to open Piksi on %s", port.c_str( ) );
	else
		ROS_INFO( "Piksi opened successfully on %s", port.c_str( ) );

	ros::spin( );

	std::exit( EXIT_SUCCESS );
}


swiftnav_piksi/include/swiftnav_piksi/piksi.h
/***************************************************************************//**
* \file piksi.h
*
* \brief Standalone C Driver for the Swift Navigation Piksi GPS (header)
* \author Scott K Logan
* \author Caleb Jamison
* \date Febuary 23, 2014
*
* API for the standalone C driver
*
* \section license License (BSD-3)
* Copyright (c) 2013, Scott K Logan\n
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* - Redistributions of source code must retain the above copyright notice,
* this list of conditions and the following disclaimer.
* - Redistributions in binary form must reproduce the above copyright notice,
* this list of conditions and the following disclaimer in the documentation
* and/or other materials provided with the distribution.
* - Neither the name of Willow Garage, Inc. nor the names of its contributors
* may be used to endorse or promote products derived from this software
* without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
******************************************************************************/

#ifndef _piksi_h
#define _piksi_h

#ifdef __cplusplus
extern "C" {
#endif

#include <libsbp/sbp.h>

enum piksi_error
{
	PIKSI_SUCCESS = 0,
	PIKSI_ERROR_IO = -1,
	PIKSI_ERROR_NO_DEVICE = -4,
	PIKSI_ERROR_NO_MEM = -11,
	PIKSI_ERROR_OTHER = -13,
	PIKSI_ERROR_TIMEOUT = -14
};

int piksi_open( const char *port, int baud );
void piksi_close( const int8_t piksid );

u8 piksi_spin( const int8_t piksid );

u32 send_cmd( u8 *data, u32 num_bytes, void* context );
u32 read_data( u8 *data, u32 num_bytes, void* context );


#ifdef __cplusplus
}
#endif

#endif /* _piksi_h */


swiftnav_piksi/include/swiftnav_piksi/piksi_driver.h
/***************************************************************************//**
* \file piksi_driver.hpp
*
* \brief ROS Implementation of the C Driver (header)
* \author Scott K Logan
* \author Caleb Jamison
* \date February 23, 2014
*
* API for the ROS driver
*
* \section license License (BSD-3)
* Copyright (c) 2013, Scott K Logan\n
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* - Redistributions of source code must retain the above copyright notice,
* this list of conditions and the following disclaimer.
* - Redistributions in binary form must reproduce the above copyright notice,
* this list of conditions and the following disclaimer in the documentation
* and/or other materials provided with the distribution.
* - Neither the name of Willow Garage, Inc. nor the names of its contributors
* may be used to endorse or promote products derived from this software
* without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
******************************************************************************/

#ifndef _piksi_driver_hpp
#define _piksi_driver_hpp

#include "swiftnav_piksi/piksi.h"

#include <libsbp/sbp.h>

#include <ros/ros.h>
#include <ros/rate.h>
#include <tf/tf.h>
#include <nav_msgs/Odometry.h>
#include <gps_common/GPSFix.h>

#include <boost/thread.hpp>

namespace swiftnav_piksi
{
  void heartbeat_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  void time_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  void pos_llh_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  void dops_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  void baseline_ned_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  void vel_ned_callback(u16 sender_id, u8 len, u8 msg[], void *context);

  class Piksi
  {
  public:
    Piksi(const ros::NodeHandle& nh);
    ~Piksi();
    bool PiksiOpen();
    void PiksiClose();

  private:
    bool PiksiOpenNoLock();
    void PiksiCloseNoLock();
    std::string GetFixDescription(uint8_t fix_type);
    void spin();
    void spinOnce();
    double GetUtcTime(uint32_t ms);

    ros::NodeHandle nh_;
    ros::NodeHandle pnh_;
    std::string port_;
    int baud_;
    std::string frame_id_;
    int8_t piksid_;
    int32_t utc_offset_;
    boost::mutex cmd_lock;

    sbp_state_t state_;
    sbp_msg_callbacks_node_t heartbeat_callback_node;
    sbp_msg_callbacks_node_t time_callback_node;
    sbp_msg_callbacks_node_t pos_llh_callback_node;
    sbp_msg_callbacks_node_t dops_callback_node;
    sbp_msg_callbacks_node_t baseline_ned_callback_node;
    sbp_msg_callbacks_node_t vel_ned_callback_node;

    ros::Publisher dop_pub_;
    ros::Publisher llh_pub_;
    ros::Publisher rtk_pub_;
    ros::Publisher time_pub_;
    ros::Publisher vel_pub_;
    ros::Publisher gps_fix_pub_;

    gps_common::GPSFix last_gps_fix_;

    ros::Rate spin_rate_;
    boost::thread spin_thread_;

    uint32_t open_failure_count_;

    friend void heartbeat_callback(u16 sender_id, u8 len, u8 msg[], void *context);
    friend void time_callback(u16 sender_id, u8 len, u8 msg[], void *context);
    friend void pos_llh_callback(u16 sender_id, u8 len, u8 msg[], void *context);
    friend void dops_callback(u16 sender_id, u8 len, u8 msg[], void *context);
    friend void baseline_ned_callback(u16 sender_id, u8 len, u8 msg[], void *context);
    friend void vel_ned_callback(u16 sender_id, u8 len, u8 msg[], void *context);
  };
}

#endif /* _piksi_driver_hpp */


swiftnav_piksi/src/piksi_driver.cpp
#include "swiftnav_piksi/piksi_driver.h"
#include <libsbp/system.h>
#include <libsbp/navigation.h>

#include <swiftnav_piksi_msgs/SbpBaseline.h>
#include <swiftnav_piksi_msgs/SbpDops.h>
#include <swiftnav_piksi_msgs/SbpGpsTime.h>
#include <swiftnav_piksi_msgs/SbpPosLlh.h>
#include <swiftnav_piksi_msgs/SbpVelNed.h>
#include <gps_common/GPSFix.h>

#include <ros/ros.h>
#include <ros/time.h>
#include <tf/tf.h>

namespace swiftnav_piksi
{
  Piksi::Piksi(const ros::NodeHandle& nh):
    nh_(nh),
    piksid_(-1),
    utc_offset_(-18), // As of 2016-12-31
    spin_rate_(2000),
    open_failure_count_(0)
  {
    pnh_ = ros::NodeHandle("~");
    pnh_.param("port", port_, std::string("/dev/ttyUSB0"));
    pnh_.param("baud", baud_, 152000);
    pnh_.param("frame_id", frame_id_, std::string("piksi_gps"));
    
    dop_pub_ = nh_.advertise<swiftnav_piksi_msgs::SbpDops>("piksi/dops", 2);
    llh_pub_ = nh_.advertise<swiftnav_piksi_msgs::SbpPosLlh>("piksi/pos_llh", 2);
    rtk_pub_ = nh_.advertise<swiftnav_piksi_msgs::SbpBaseline>("piksi/baseline", 2);
    time_pub_ = nh_.advertise<swiftnav_piksi_msgs::SbpGpsTime>("piksi/time", 2);
    vel_pub_ = nh_.advertise<swiftnav_piksi_msgs::SbpVelNed>("piksi/vel", 2);

    gps_fix_pub_ = nh_.advertise<gps_common::GPSFix>("piksi/gps", 2);

    ROS_INFO("Opening Piksi on %s at %d baud", port_.c_str(), baud_);
    if(!PiksiOpen())
    {
      ROS_ERROR( "Failed to open Piksi on %s at %d baud", port_.c_str(), baud_);
      ros::shutdown();
      return;
    }
    else
    {
      ROS_INFO( "Piksi opened successfully on %s at %d baud", port_.c_str(), baud_);
    }

    // If we reach this point, we should be able to start the spin thread
    boost::thread t(boost::bind(&Piksi::spin, this));
    spin_thread_.swap(t);
  }

  Piksi::~Piksi()
  {
    spin_thread_.interrupt();
    PiksiClose();
  }

  bool Piksi::PiksiOpen()
  {
    boost::mutex::scoped_lock lock(cmd_lock);
    return PiksiOpenNoLock();
  }

  bool Piksi::PiksiOpenNoLock()
  {
    if (piksid_ >= 0)
    {
      return true;
    }

    piksid_ = piksi_open(port_.c_str(), baud_);

    if (piksid_ < 0)
    {
      open_failure_count_++;
      return false;
    }

    sbp_state_init(&state_);
    sbp_state_set_io_context(&state_, &piksid_);

    sbp_register_callback(&state_, SBP_MSG_HEARTBEAT,    &heartbeat_callback,    (void*) this, &heartbeat_callback_node);
    sbp_register_callback(&state_, SBP_MSG_GPS_TIME,     &time_callback,         (void*) this, &time_callback_node);
    sbp_register_callback(&state_, SBP_MSG_POS_LLH,      &pos_llh_callback,      (void*) this, &pos_llh_callback_node);
    sbp_register_callback(&state_, SBP_MSG_DOPS,         &dops_callback,         (void*) this, &dops_callback_node);
    sbp_register_callback(&state_, SBP_MSG_BASELINE_NED, &baseline_ned_callback, (void*) this, &baseline_ned_callback_node);
    sbp_register_callback(&state_, SBP_MSG_VEL_NED,      &vel_ned_callback,      (void*) this, &vel_ned_callback_node);
    
    return true;
  }

  void Piksi::PiksiClose()
  {
    boost::mutex::scoped_lock lock(cmd_lock);
    PiksiCloseNoLock();
  }

  void Piksi::PiksiCloseNoLock()
  {
    int8_t old_piksid = piksid_;
    if (piksid_ < 0)
    {
      return;
    }
    piksid_ = -1;
    piksi_close(old_piksid);

    if (llh_pub_)
    {
      llh_pub_.shutdown();
    }

    if (time_pub_)
    {
      time_pub_.shutdown();
    }
  }

  void heartbeat_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting heartbeat callback.");
      return;
    }

    msg_heartbeat_t hb = *(msg_heartbeat_t*) msg;

    class Piksi* driver = (class Piksi*) context;
    
    // Nothing to do right now.

    return;
  }

  void time_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting Time callback.");
      return;
    }

    class Piksi* driver = (class Piksi*) context;

    msg_gps_time_t time = *(msg_gps_time_t*) msg;

    swiftnav_piksi_msgs::SbpGpsTimePtr time_msg( new swiftnav_piksi_msgs::SbpGpsTime );

    time_msg->header.frame_id = driver->frame_id_;
    time_msg->header.stamp = ros::Time::now( );
    
    time_msg->weeks = time.wn;
    time_msg->time_of_week = time.tow;
    time_msg->nano = time.ns;
    time_msg->time_source = (time.flags & 0x0007);
    std::string desc("None");
    if (time_msg->time_source)
    {
      desc = std::string("GNSS");
    }
    time_msg->time_source_description = desc;

    driver->time_pub_.publish(time_msg);

    return;
  }

  void pos_llh_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting pos_llh callback.");
      return;
    }

    class Piksi* driver = (class Piksi*) context;

    msg_pos_llh_t llh = *(msg_pos_llh_t*) msg;

    swiftnav_piksi_msgs::SbpPosLlhPtr llh_msg( new swiftnav_piksi_msgs::SbpPosLlh );

    llh_msg->header.frame_id = driver->frame_id_;
    llh_msg->header.stamp = ros::Time::now( );

    llh_msg->time_of_week = llh.tow;

    llh_msg->latitude = llh.lat;
    llh_msg->longitude = llh.lon;
    llh_msg->altitude = llh.height;

    llh_msg->h_accuracy = static_cast<double>(llh.h_accuracy) * 0.001;
    llh_msg->v_accuracy = static_cast<double>(llh.v_accuracy) * 0.001;

    llh_msg->sats = llh.n_sats;

    llh_msg->raim_repair = ((llh.flags & 0x0080) >> 7);
    llh_msg->fix_type = (llh.flags & 0x007);
    llh_msg->fix_description = driver->GetFixDescription(llh_msg->fix_type);

    driver->llh_pub_.publish(llh_msg);

    // TODO(kkozak): Synchronize the consituent messages for this GPS fix message.
    driver->last_gps_fix_.header = llh_msg->header;

    // TODO(kkozak): Check to see if we can distinguish between sats visible
    //               and used.
    driver->last_gps_fix_.status.satellites_used = llh.n_sats;
    driver->last_gps_fix_.status.satellites_visible = llh.n_sats;
    // TODO(kkozak): Set measurement status (i.e. rtk status)
    
    // TODO(kkozak): Fix time -- the commented method below is not correct.
    driver->last_gps_fix_.time = driver->GetUtcTime(llh.tow);
    driver->last_gps_fix_.latitude = llh.lat;
    driver->last_gps_fix_.longitude = llh.lon;
    driver->last_gps_fix_.altitude = llh.height;

    driver->last_gps_fix_.err_horz = llh_msg->h_accuracy;
    driver->last_gps_fix_.err_vert = llh_msg->v_accuracy;
    driver->last_gps_fix_.position_covariance[0] = std::pow(llh_msg->h_accuracy, 2);
    driver->last_gps_fix_.position_covariance[4] = std::pow(llh_msg->h_accuracy, 2);
    driver->last_gps_fix_.position_covariance[8] = std::pow(llh_msg->v_accuracy, 2);
    driver->last_gps_fix_.position_covariance_type = 2;

    driver->gps_fix_pub_.publish(driver->last_gps_fix_);
    return;
  }

  void dops_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting dops_callback.");
      return;
    }
		
    msg_dops_t dops = *(msg_dops_t*) msg;

    class Piksi* driver = (class Piksi*) context;
    swiftnav_piksi_msgs::SbpDopsPtr dop_msg(new swiftnav_piksi_msgs::SbpDops);
    dop_msg->header.frame_id = driver->frame_id_;
    dop_msg->header.stamp = ros::Time::now();

    dop_msg->time_of_week = dops.tow;
    dop_msg->gdop = static_cast<double>(dops.gdop) * 0.01;
    dop_msg->pdop = static_cast<double>(dops.pdop) * 0.01;
    dop_msg->tdop = static_cast<double>(dops.tdop) * 0.01;
    dop_msg->hdop = static_cast<double>(dops.hdop) * 0.01;
    dop_msg->vdop = static_cast<double>(dops.vdop) * 0.01;

    driver->dop_pub_.publish(dop_msg);

    // TODO(kkozak): The spec shows the dops message as having flags, but that 
    //               field doesn't seem to exist with the version we're using
    // dop_msg->raim_repair = ((dops.flags & 0x0080) >> 7);
    // dop_msg->fix_type = (dops.flags & 0x0007);
    // dop_msg->fix_description = driver->GetFixDescription(dop_msg->fix_type);
    driver->last_gps_fix_.gdop = dop_msg->gdop;
    driver->last_gps_fix_.pdop = dop_msg->pdop;
    driver->last_gps_fix_.tdop = dop_msg->tdop;
    driver->last_gps_fix_.hdop = dop_msg->hdop;
    driver->last_gps_fix_.vdop = dop_msg->vdop;

    // TODO(kkozak): Publish dop_msg? 
    return;
  }

  void baseline_ned_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting baseline_ned_callback.");
      return;
    }

    class Piksi* driver = (class Piksi*) context;

    msg_baseline_ned_t sbp_ned = *(msg_baseline_ned_t*) msg;

    swiftnav_piksi_msgs::SbpBaselinePtr baseline_msg( new swiftnav_piksi_msgs::SbpBaseline );

    baseline_msg->header.frame_id = driver->frame_id_;
    // For best accuracy, header.stamp should maybe get tow converted to ros::Time
    baseline_msg->header.stamp = ros::Time::now( );

    baseline_msg->time_of_week = sbp_ned.tow;
    baseline_msg->north = static_cast<double>(sbp_ned.n) * 0.001;
    baseline_msg->east = static_cast<double>(sbp_ned.e) * 0.001;
    baseline_msg->down = static_cast<double>(sbp_ned.d) * 0.001;
    baseline_msg->h_accuracy = static_cast<double>(sbp_ned.h_accuracy) * 0.001;
    baseline_msg->v_accuracy = static_cast<double>(sbp_ned.v_accuracy) * 0.001;
    baseline_msg->sats = sbp_ned.n_sats;
    baseline_msg->raim_repair = ((sbp_ned.flags & 0x0080) >> 7);
    baseline_msg->fix_type = (sbp_ned.flags & 0x0007);
    baseline_msg->fix_description = driver->GetFixDescription(baseline_msg->fix_type);

    driver->rtk_pub_.publish(baseline_msg);
    return;
  }

  void vel_ned_callback(u16 sender_id, u8 len, u8 msg[], void* context)
  {
    if (context == NULL)
    {
      ROS_ERROR("Context void. Exiting vel_ned_callback");
      return;
    }

    class Piksi* driver = (class Piksi*) context;

    msg_vel_ned_t sbp_vel = *(msg_vel_ned_t*) msg;

    swiftnav_piksi_msgs::SbpVelNedPtr vel_msg(new swiftnav_piksi_msgs::SbpVelNed);
    vel_msg->header.frame_id = driver->frame_id_;
    vel_msg->header.stamp = ros::Time::now();

    vel_msg->time_of_week = sbp_vel.tow;
    vel_msg->vel_north = static_cast<double>(sbp_vel.n) * 0.001;
    vel_msg->vel_east = static_cast<double>(sbp_vel.e) * 0.001;
    vel_msg->vel_down = static_cast<double>(sbp_vel.d) * 0.001;
    vel_msg->h_accuracy = static_cast<double>(sbp_vel.h_accuracy) * 0.001;
    vel_msg->v_accuracy = static_cast<double>(sbp_vel.v_accuracy) * 0.001;
    
    vel_msg->sats = sbp_vel.n_sats;
    vel_msg->velocity_mode = (sbp_vel.flags & 0x0007);
    std::string desc("Invalid");
    if (vel_msg->velocity_mode == 1)
    {
      desc = std::string("Measured Doppler derived");
    }
    else if (vel_msg->velocity_mode == 2)
    {
      desc = std::string("Computed Doppler derived");
    }
    vel_msg->velocity_mode_description = desc;

    driver->vel_pub_.publish(vel_msg);

    // TODO(kkozak): We're using this velocity message to infer heading for now,
    //               but the piksi can output a heading message that can 
    //               possibly be used instead, though it may only give relative 
    //               direction of rover with respect to base, which is not the 
    //               heading that we care about.
    double vx = vel_msg->vel_north;
    double vy = vel_msg->vel_east;
    double speed = std::sqrt(vx*vx + vy*vy);
    driver->last_gps_fix_.speed = speed;
    driver->last_gps_fix_.track =  std::atan2(vy, vx) * 180.0 / M_PI;
    
    return;
  }


  double Piksi::GetUtcTime(uint32_t ms)
  {
    // TODO(kkozak): This needs to be verified, and offset needs to be captured
    //               automatically, rather than from the default value.
    // 2017-03-03 -- It looks right compared to online UTC clock
    double seconds = static_cast<double>(ms) * 0.001 + utc_offset_;
    int32_t days = seconds / 86400.0;
    double utc_time = seconds - days * 86400.0;
    return utc_time;
  }

  std::string Piksi::GetFixDescription(uint8_t fix_type)
  {
    std::string fix_description("");
    switch (fix_type)
    {
    case 0:
      fix_description = std::string("Invalid");
      break;
    case 1:
      fix_description = std::string("SPP");
      break;
    case 2:
      fix_description = std::string("DGNSS");
      break;
    case 3:
      fix_description = std::string("Float RTK");
      break;
    case 4:
      fix_description = std::string("Fixed RTK");
      break;
    default: 
      fix_description = "Invalid";
    }
    return fix_description;
  }
  void Piksi::spin()
  {
    while (ros::ok())
    {
      boost::this_thread::interruption_point();
      Piksi::spinOnce();
      spin_rate_.sleep();
    }
  }

  void Piksi::spinOnce()
  {
    int ret;

    cmd_lock.lock();
    if(piksid_ < 0 && !PiksiOpenNoLock())
    {
      cmd_lock.unlock();
      return;
    }

    ret = sbp_process( &state_, &read_data );
    cmd_lock.unlock();
  }
}
